### PR TITLE
RSP-1861 [Docs][A11y] Manage focus when breadcrumb changes in link Popover

### PR DESCRIPTION
**Pull request by majornista**
When a link within a popover changes the contents of the popover, focus will be lost to the document.body, because the element that had focus will be removed from the DOM and replaced with new content.  We should manage focus by explicitly setting focus to the current breadcrumb item when the content changes.

Closes [RSP-1861](https://jira.corp.adobe.com/browse/RSP-1861)

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [Issue RSP-1861](https://jira.corp.adobe.com/browse/RSP-1861)
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. yarn start:docs
2. Navigate to http://localhost:1234/react-aria/useListBox.html
3. Click on a link that opens a popover displaying information about the element type, like "AriaListBoxOptions".
4. Tab to another link to a React-Spectrum object type like "CollectionChildren".
5. Press Enter to activate the link, updating the contents of the Popover.
6. Focus will be set to the current breadcrumb, but since it is static content, no focus ring is displayed. Think of it like an in-page anchor link. 
7. Press Tab or Shift+Tab to navigate to the next or previous interactive control from the current focus in the tab navigation order.
8. The focus should move to the first focusable link after the Breadcrumbs in the Popover content or to the previous Breadcrumb Item from the current Breadcrumb

## 🧢 Your Project:

Accessibility

### CODE REVIEW

**Comment by soulhi9988** [comment]
[Image.jpeg]

ดาวน์โหลด Outlook for iOS<https://aka.ms/o0ukef>
________________________________
จาก: adobe-bot <notifications@github.com>
ส่ง: Monday, July 20, 2020 10:15:54 PM
ถึง: adobe/react-spectrum <react-spectrum@noreply.github.com>
สำเนาถึง: Subscribed <subscribed@noreply.github.com>
ชื่อเรื่อง: Re: [adobe/react-spectrum] RSP-1861 [Docs][A11y] Manage focus when breadcrumb changes in link Popover (#728)


Build successful! 🎉

  *   View the storybook<https://reactspectrum.blob.core.windows.net/reactspectrum/e640f1ce872f170df179dcb8f1af5272f92e5945/storybook/index.html>
  *   View the documentation<https://reactspectrum.blob.core.windows.net/reactspectrum/e640f1ce872f170df179dcb8f1af5272f92e5945/docs/index.html>

File: packages/dev/docs/src/docs.js
Comment (snowystinger) [inline packages/dev/docs/src/docs.js:88]: the actual cancelling should take place in the useEffect cleanup, isCancelled will be a different instance in the next run of the useEffect, so I don't really see how this is currently doing anything

File: packages/dev/docs/src/docs.js
Comment (snowystinger) [inline packages/dev/docs/src/docs.js:65]: I do not understand what is happening here, why is aria-haspopup removed?

**Review by snowystinger — COMMENTED** [review]
I see the issue and i see that it's doing the expected thing after the change
I don't follow the css changes in here though, can you explain more about them?
It seems like the Breadcrumbs component should be handling some of this?

File: packages/dev/docs/src/docs.js
Comment (majornista) [reply packages/dev/docs/src/docs.js:65]: When a `Link` within a code sample appears within a `LinkPopover` it will no longer open in a new popover, but instead refreshes the contents of the current popover, so `aria-haspopup` is not really appropriate.

File: packages/dev/docs/src/docs.js
Comment (majornista) [reply packages/dev/docs/src/docs.js:88]: Yeah, it's probably not doing anything, and I'm not sure that it's needed.

Comment (majornista) [reply packages/dev/docs/src/docs.js]: Should be fixed now.

**Comment by majornista** [comment]
> I see the issue and i see that it's doing the expected thing after the change
> I don't follow the css changes in here though, can you explain more about them?
> It seems like the Breadcrumbs component should be handling some of this?

Normally, the current Breadcrumb does not receive keyboard focus, but with this code documentation popover, we have to make sure that the focus isn't lost to the document.body when the popover contents change should a user drill down within the documentation. For this reason, when the Breadcrumbs update, I make the current BreadcrumbItem focusable and set focus to it, but when it blurs, we restore the expected behavior for Breadcrumbs, that the current BreadcrumbItem does not receive focus. Think of it like jumping to an in page link.

File: packages/dev/docs/src/docs.js
Comment (snowystinger) [inline packages/dev/docs/src/docs.js:92]: i still do not follow the logic going on here
is this really two useEffects?

or a useLayoutEffect and a useEffect?

at the very least, this return needs to actually cancel the setTimeout so it can't be called on an unmounted component where it can't access anything because it's no longer in the dom

Comment (majornista) [reply packages/dev/docs/src/docs.js]: Ok, I think I have it now: https://github.com/adobe/react-spectrum/pull/728/commits/6526bb0f113a05366cfbaeb648ee56b928ecb5f3#diff-27e06259cf9af05851fc3cc84ecc75120dcbf242051ad4acc6bfa09d892f8b07R61-R96

File: packages/@adobe/spectrum-css-temp/components/breadcrumb/index.css
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/breadcrumb/index.css:137]: this block is the same as above, lets keep them combined
```
.spectrum-Breadcrumbs-item.is-dragged .spectrum-Breadcrumbs-itemLink:before,
.spectrum-Breadcrumbs-itemLink:not([aria-current]):focus-ring:before {
```

File: packages/@adobe/spectrum-css-temp/components/breadcrumb/index.css
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/breadcrumb/index.css:119]:
```suggestion
.spectrum-Breadcrumbs-item.is-dragged .spectrum-Breadcrumbs-itemLink:before,
```

**Review by snowystinger — APPROVED** [review]
LGTM, tested on Chrome/Safari/FF

**Review by LFDanLu — APPROVED** [review]
LGTM, tested in FF, Chrome, and Safari. @majornista I'll let you merge this if you think it is good to go in

**Comment by majornista** [comment]
Don't think I'm authorized to merge.

**Comment by LFDanLu** [comment]
Ah, no problem I'll merge then

**Pull request merged**
